### ISSUE TITLE
Add replaceTrack method to MediaStream

### ISSUE THREAD

**Issue by guest271314** [w3c/mediacapture-record#167]
Related:

- https://github.com/w3c/mediacapture-record/issues/4

- https://github.com/w3c/mediacapture-record/issues/166

- https://github.com/w3c/mediacapture-main/issues/575

- https://github.com/w3c/webrtc-pc/issues/2171

Given a `MediaStream` instance either created using `new MediaStream()` and/or by `getUserMedia()`, `HTMLMediaElement.captureStream()`, add a `replaceTrack()` method directly to the existing `MediaStream` which behaves similarly to `RTCRtpSender.replaceTrack()` https://w3c.github.io/webrtc-pc/#dom-rtcrtpsender-replacetrack

> NOTE
> 
> There is not an exact 1:1 correspondence between tracks sent by one RTCPeerConnection and received by the other. For one, IDs of tracks sent have no mapping to the IDs of tracks received. Also, replaceTrack changes the track sent by an RTCRtpSender without creating a new track on the receiver side; the corresponding RTCRtpReceiver will only have a single track, potentially representing **multiple sources of media stitched together**. Both addTransceiver and replaceTrack can be used to cause the same track to be sent multiple times, which will be observed on the receiver side as multiple receivers each with its own separate track. Thus it's more accurate to think of a 1:1 relationship between an RTCRtpSender on one side and an RTCRtpReceiver's track on the other side, matching senders and receivers using the RTCRtpTransceiver's mid if necessary. (emphasis added)

The primary use case for adding a `replaceTrack` method to `MediaStream` is for the ability to record media (for example following a call to `HTMLMediaElement.captureStream()`) with `MediaRecorder` where the `src` attribute of the `HTMLMediaElement` is changed (currently `MediaRecorder` stops recording when a track is added to the `MediaStream` being recorded by means of `src` of `<video>` or `<audio>` element changing). 

This https://github.com/guest271314/MediaFragmentRecorder/blob/webrtc-replacetrack/MediaFragmentRecorder.html is currently possible using `RTCPeerConnection` (proof of concept) though should be possible without the end-user having to manually create two `RTCPeerConnection` instances to achieve the requirement. Instead, the end-user should simply be able to execute `<MediaStreamInstance>.replaceTrack(withTrack)` where internally `replaceTrack(withTrack)` achieves the same functionality of `RTCRtpSender.replaceTrack()` 

> multiple sources of media stitched together

> which will be observed on the receiver side as multiple receivers each with its own separate track

Filing this issue before filing a PR for changes to the respective specification(s).

**Comment by alvestrand** [comment]
This seems like the right repo for it, since the justification for the function is strictly based on the MediaRecorder definition that "adding a track stops recording".
An alternative would be to let MediaRecorder do the same thing as the &lt;video&gt; playback, and let only one video track be recorded, but the video track can be replaced by add/remove.

**Comment by guest271314** [comment]
@jan-ivar Any progress of composing the specification portions for a `replaceTrack()` method for `MediaRecorder`?

**Comment by henbos** [comment]
I was going to propose replaceStream on the media recorder.
@jan-ivar was going to propose detecting if removeTrack+addTrack happened in the same task then the recorded could detect that and act as if it was replaced.

**Comment by guest271314** [comment]
@henbos Conceivably and programmatically there is more than one way to achieve the expected result for the use case of recording multiple tracks potentially having differing resolutions. 

Note, ideally, when the track change occurs, either via replaceStream or addTrack+removeTrack without `MediaRecorder` stopping, it should be specified that the `width` and `height` of the new video track be checked and set as the current encoded width and height (e.g., see https://bugs.chromium.org/p/chromium/issues/detail?id=972470; https://github.com/w3c/mediacapture-record/pull/172; https://bugs.chromium.org/p/webm/issues/detail?id=1642), not to automatically encode the new track frames as the initial input video track width and height, to avoid potential for confusion as to whether or not the replacement track pixel dimensions, resolution, width, height should be constrained to the initial track encoding parameters, which the encoding of the replacement video track should not, unless explicitly set as a constant width and height by the user.

**Comment by guest271314** [comment]
@henbos How does the replaceStream approach intend to handle the case of replacing a stream that has AV with a stream that has only V or A?

@jan-ivar Could addTrack+removeTrack be extended to handle to case of A track sample rate 44100 being substituted with A track sample rate 48000?

See https://www.reddit.com/r/mkvtoolnix/comments/cdi824/mkvmerge_prints_error_when_merging_webm_files/etwhugs/.

Or, are those interoperability concerns beyond the scope of the `MediaRecorder` specification?

**Comment by guest271314** [comment]
Note, the origin of this issue is the code at https://github.com/guest271314/MediaFragmentRecorder/blob/master/MediaFragmentRecorder.html which outputs the expected result of a WebM video which encodes the individual track pixel dimensions at Mozilla Firefox, does not output the expected result at Chromium, Chrome browser. Would suggest to run the code to observe the output at the respective browsers. That lead to attempting to compose code which output the same result at each browser https://github.com/guest271314/MediaFragmentRecorder/branches, observing several different issues at each browser due to implementations of APIs, methods during testing code.

Was able to encode variable resolution at Chromium browser using the codec `"video/<x-matroska|webm>;codecs=h264"` or `"video/<x-matroska|webm>;codecs=avc1"` https://plnkr.co/edit/Axkb8s?p=info. Even when the Matroska file is created, Chromium has the unexpected bug that navigating to a Matroska (`.mkv`) file prompts user to download the file instead of loading the file and playing the file at HTML `<video>` element, requiring the workaround of saving the Matroska file with `.webm` extension, which leads to the issue of a technically "invalid" WebM file, see https://bugs.chromium.org/p/chromium/issues/detail?id=999580, https://bugs.chromium.org/p/chromium/issues/detail?id=980822. Mozilla Firefox does not playback the video, for reasons described in issues relating to codecs in WebM and Matroska file output by `MediaRecorder` 
at Firefox and Chromium (see last link at previous sentence). Mozilla Firefox already does output expected result using that browsers' implementation of VP8, VP9, and related encoders and decoders and HTML `<video>` element. 

Without using `"video/<x-matroska|webm>;codecs=avc1"` `mimeType` at `MediaRecorder` at Chromium have not been able to playback WebM files containing variable resolution which output the expected pixel dimensions at HTML `<video>` element, which for reasons unknown here, does output variable width and height when the codec is `h264` or `avc1`.  

The purpose of this issue is not to simply have methods for each browser that still output dissimilar results, for example, arbitrary track order, which require more than one version of code to achieve similar output.

The purpose of this issue is for the capability to compose the same code for each browser and achieve the same output at each browser, to the extent possible, uniform implementation of any extended method specification which is intended to not stop `MediaRecorder` recording and encoding replacement track properties (for example, width and height) when replacing, adding and removing tracks and streams at `MediaRecorder`.

**Comment by henbos** [comment]
> it should be specified that the width and height of the new video track be checked and set as the current encoded width and height

That may make sense to avoid surprises, a good thing for the PR to resolve this issue. However, tracks can already change resolution on the fly (example: WebRTC remote tracks). What to do if resolution changes seems tangental to this issue, and I think should be handled in a separate issue. This spec currently seems a bit underspecified.

> @henbos How does the replaceStream approach intend to handle the case of replacing a stream that has AV with a stream that has only V or A?

In the spirit of existing text to stop recording if tracks are removed, I think the operation should fail if you attempt to replace it by a stream that does not have the same number of audio and video tracks.

I also wonder what to do if there are multiple tracks of the same kind, since the order of tracks is undefined and what does it mean to record multiple tracks when the output is a blob? But I think this is another tangent to the issue in question of being able to replace tracks.

**Comment by henbos** [comment]
I created a PR but I didn't mention anything about resolution changing for now

**Comment by jan-ivar** [comment]
I've been pondering whether this is going to work right with resolution changes. Does Chrome let you record a remote track that changes resolution? AFAIK Firefox doesn't allow recording of remote tracks period.

I also did this [demo page](https://jan-ivar.github.io/dummy/replace-track.html) where I remove a stream's video track and add a different video track on the same event loop (the API alternative I mentioned). This causes `UnknownError` from MediaRecorder in Firefox. In Chrome there's no error, but I don't see a recording (maybe autoplay restrictions interfering?) cc @Pehrsons 

Just wanted to share what I'd tried.

**Comment by guest271314** [comment]
@henbos 

Chromium, Chrome has specific issues relevant to handling width and height resolution changes during mid-stream this issue and linked issues cover some of the crashes and unexpected behaviour of Chromium implementation of encoding and decoding depending on the approach used, e.g., https://bugs.chromium.org/p/chromium/issues/detail?id=1001948. Concur, separate issue though related as the expected result is to merge multiple `MediaStreamTrack`s having potential variable resolution. In Chromium the output could be far different from the input. When the specification mandates an exact output, the definition of what is expected is clear, unambiguous, not possible to convert to a different defintion at implementation. Mozilla Firefox and Nightly do not have issues recording or displaying variable resolution video.

No, the operation should not fail if the exact number of corresponding tracks exist in the next stream. The developer has the capability to create both video and audio `MediaStreamTrack` to add to the `MediaStream` before `MediaRecorder` `start()` is executed. The behaviour of stopping recording when tracks are added or removed is not the expected behaviour which this was filed for, the issue was filed to fix that behaviour, e.g., https://github.com/guest271314/native-messaging-mkvmerge/blob/master/app/native-messaging-mkvmerge.js#L81, https://github.com/guest271314/native-messaging-mkvmerge/blob/master/app/native-messaging-mkvmerge.js#L121, stream video frames or audio frames to complement the corresponding <A|V> track. Track order is another issue that is related. 

Replacement of stream is the core. From there adjustments can be made where the adjacent settings and options and crashes for certain content is duly noted.

**Comment by guest271314** [comment]
@jan-ivar Yes, Chromium provides means to record variable resolution VP8, h264, and avc1 codecs. Playback of the variable resolution WebM video at Chromium is a separate issue. See the link to Chromium bug at previous post, see also links to other issues. In brief, Chromium implementation of `MediaRecorder` VP8 codecs and HTML `<video>` element does not display the exact input frame resolution for each frame. `video/x-matroska;codecs=h264` or `video/webm;codecs=avc1` or vice versa do display variable resolution output at HTML `<video>` element https://plnkr.co/edit/Axkb8s?p=info, though do to the codecs used is not currently playable at Mozilla browsers. Mozilla Firefox and Nightly do not have any issue recording an HTML `<video>` element with dynamic resolution changes and outputting exactly if not precisely each frame that was recorded at `<video>` element. Chromium crashes the browser when executing `captureStream()` on a `<video>` element when resolution changes at the first `resize` event. 

In fact, video recorded at Mozilla does resize HTML `<video>` element at Chromium. This version of code using `MediaRecorder` does output variable resolution frames at Mozilla browers https://github.com/guest271314/MediaFragmentRecorder/tree/imagecapture-audiocontext-readablestream-writablestream. video recorded at Chromium However, the two implementations of encoding WebM are distinct enough from each other to not be capable of being merged. That is what the gist of this issue is about: writing code which outputs the same result at Chromium and Firefox. 

If not for the initial and maintained use case of composing the same code for both browsers would have stopped at version 0 at Firefox using `MediaSource`. However, that is not the case https://github.com/guest271314/MediaFragmentRecorder/branches. Yes, it is possible to record "remote" WebRTC tracks via `MediaStream` at Firefox, Nightly see the "master" and "webtrc-replacetrack" (and "canvas") branches at the linked repository.

To an appreciable already solved the issue as to what the expected output is with workarounds. That is, observing and noting and working around various bugs, primarily Chromium browser crashes and Firefox clipping or muting the last approximately 1 second of audio. Each branch contains links the bugs found while experimenting. Not certain how many there were, though more than likely filed the majority. The workarounds are insuffient. The "webtrc-replacetrack" approach at the linked repository 

The implementation of events, constraints, `getSettings()`, etc. relating to `MediaStream` and `MediaStreamTrack` at Firefox and Chromium are not exactly the same, e.g., see https://github.com/w3c/mediacapture-fromelement/issues/78.

The specification is the goal. Meaning, note each point that crashed the browser needs to be properly addressed equally to the overall expected output of replaced tracks and streams. That means not omitting thoroughly defining track order, resolution changes, muting of last secod of audio.

**Comment by guest271314** [comment]
@henbos Since Media Capture and Stream does not define any order for `MediaStreamTrack`s within a `MediaStream` the order is arbitrary both between instances of `MediaRecorder` at the same browser and between Firefox and Chromium, Chrome. There must be some form of algorithm to replace the correct track in the existing media stream. This is the second version of the algorithm for merging WebM tracks output by `MediaRecorder` at Chromium after having created the corresponding A or V track if not encoded in the original stream, e.g, [A0] (input) = [A0,V0] (create V0 for input A0) + [V1,A1] (append V1 to V0, A1 to V0 note arbitrary track order) + [V2] (input) => [A2, V2] (create audio track for V2) => [A_0, V_0] // output

  ```
  // https://www.reddit.com/r/mkvtoolnix/comments/cdi824/mkvmerge_prints_error_when_merging_webm_files/etwhugs/
    // 1. run `mkvmerge -J` file.webm on each input file
    // 2. determine which tracks go together
    // 3. create an appropriate --append-to argument from the algorithm in 2
    let filesMetadata = [];
    for (const fileName of fileNames) {
      await new Promise(resolve => {
        const getMetadata = ({
          body
        }) => {
          port.onMessage.removeListener(getMetadata);
          filesMetadata.push(JSON.parse(body));
          resolve();
        };
        port.onMessage.addListener(getMetadata);
        port.postMessage({
          "message": "metadata",
          "body": `${cmd} ${metadata} ${fileName}`
        });
      });
    };
    const tracks = filesMetadata.map(({tracks}) => tracks);
    const trackList = tracks.map(track => ({audio:getTrack(track, "audio").id, video:getTrack(track, "video").id}));
    // check if tracks are ordered AV,AV...AV or arbitrarily AV,VA,AV,AV,VA...AV
    const orderedTracks = tracks.every(([{type}]) => type === "audio");
    // construct `--append-to` option for merging files where
    // tracks are not in consistent order; for example, WebM
    // files output by Chromium, Firefox MediaRecorder implementations 
    // Chromium => Opus: "id": 0, Firefox => Opus: "id": 1,
    // Chromium => VP8: "id": 1, Firefox => VP8: "id": 0 
    // merging a WebM file output by MediaRecorder at Chromium 
    // with a WebM file produced by MediaRecorder at Firefox crashes tab
    trackList.reduce((a, b, index) => (
      appendTo += `${index}:${b.audio}:${index-1}:${a.audio},${index}:${b.video}:${index-1}:${a.video}${trackList[index+1]?",":""}`
      , b
    ));
    console.log(trackList, JSON.stringify({
      filesMetadata, orderedTracks, appendTo
    }, null, 2));
    port.onMessage.addListener(onNativeMessage);
    const message = {
      "message": "write",
      // if tracks in files are not ordered use `--append-to` option, else do not
      "body": `${cmd} ${options} ${outputFileName} ${!orderedTracks ? appendTo : ""} '[' ${fileNames.join(" ")} ']'`
    };
    // mkvmerge <35 outputs error when merging video tracks having differing pixel dimensions
    // Error: The track number 1 from the file '1.webm' cannot be appended to the track number 
    // 1 from the file '0.webm'. The width of the two tracks is different: 768 and 480
    // fixed with mkvmerge: allow appending video tracks with differing pixel dimensions 
    // https://gitlab.com/mbunkus/mkvtoolnix/commit/2548a7f3497940135e55be604a87b9495fdff94d
    port.postMessage(message);
```
output https://next.plnkr.co/plunk/8J61Rw. (see https://github.com/pthatcherg/web-codecs/issues/11#issuecomment-529552845)

**Comment by henbos** [comment]
@guest271314 I get a feeling that there are multiple discussions going on in parallel, can we separate these discussions (e.g. into separate issues or a bullet point list of the different problems) so that the discussions are easier to follow?
- Bugs: If there are bugs in browsers that causes crashes or unexpected output, fixing those bugs is up to each browser implementation and not something new APIs should enable you to "work around".
- Interoperability: If there are ambiguities in the spec that allows different implementations to behave observably different causing interoperability problems, the spec needs to be clarified and the implementations updated to match.
- New features: Explanation of the bahavior proposed to be supported and a discussion about how to achieve this.

I am operating under the assumption that this is a feature request to allow changing out which audio track or video track (or both) the recorder should gather samples from, and to be able to do this in a seamless fashion (e.g. no glitches). **Is this correct?**

I think "how to handle resolution changing" (whether it changed due to remote tracks having varying resolutions or due to replaceStream() or similar APIs) should be discussed separately. We could fail or we could resize or whatever.

You also mentioned recording when the set of tracks is different in the new stream. This would be a separate use case than wanting to achieve "replaceTrack", but an API like this would achieve that too since the application can generate "silent" tracks.

As for order of tracks being unspecified and what happens if you try to record something with multiple audio or multiple video tracks: I think this should be clarified separately as well, I find the spec very unclear about this.

**Comment by henbos** [comment]
> I am operating under the assumption that this is a feature request to allow changing out which audio track or video track (or both) the recorder should gather samples from, and to be able to do this in a seamless fashion (e.g. no glitches).

See [this slide for TPAC next week](https://docs.google.com/presentation/d/1Xprz0ElNTUWRDdX6kyp11cZ4ItTvUFUmzoXsKZdXxuU/edit#slide=id.g62a86e6d8c_3_12).

**Comment by guest271314** [comment]
> I am operating under the assumption that this is a feature request to allow changing out which audio track or video track (or both) the recorder should gather samples from, and to be able to do this in a seamless fashion (e.g. no glitches). **Is this correct?**

Yes.

**Comment by Pehrsons** [comment]
> I've been pondering whether this is going to work right with resolution changes. Does Chrome let you record a remote track that changes resolution?

Tracks change resolution all the time, so there are no new problems surfacing from adding an API like this.

> AFAIK Firefox doesn't allow recording of remote tracks period.

We have allowed and supported recording remote tracks for years. [Bug 1212237](https://bugzilla.mozilla.org/show_bug.cgi?id=1212237) makes it a bit annoying though.

> I also did this [demo page](https://jan-ivar.github.io/dummy/replace-track.html) where I remove a stream's video track and add a different video track on the same event loop (the API alternative I mentioned). This causes `UnknownError` from MediaRecorder in Firefox. In Chrome there's no error, but I don't see a recording (maybe autoplay restrictions interfering?) cc @Pehrsons

Right, we don't fire InvalidModificationError for track set changes yet. I filed [bug 1581139](https://bugzilla.mozilla.org/show_bug.cgi?id=1581139).

> Just wanted to share what I'd tried.

**Comment by jan-ivar** [comment]
> We have allowed and supported recording remote tracks for years. Bug 1212237 makes it a bit annoying though.

My mistake. I was able to work around that bug and make it work: https://jsfiddle.net/jib1/ph0gad3c/

**Comment by jan-ivar** [comment]
Note that an alternative method is being discussed in https://github.com/w3c/mediacapture-record/pull/186#issuecomment-531421759

@henbos WDYT? Do we want to present both?

**Comment by guest271314** [comment]
@jan-ivar @henbos Yes, present both. What is the reason to not present both? Each approach can be useful for different cases. Is there really any decisive bright line where either prospective specification has a greater utility than the other? If there are theories as to which approach would be more useful and thus should be focused on and the other omitted from consideration, how could that result be conclusively determined without testing each for various use cases?

**Comment by alvestrand** [comment]
Closed because of lack of developer interest.

**Comment by Kaiido** [comment]
@alvestrand when you say "developer" you mean *implementer*, right? Because usually "developer" is used to talk about web-developers, and there is definitely an interest from us for such a feature.

**Comment by realies** [comment]
I would prefer not to use a `RTCPeerConnection` workaround. @alvestrand, can we reopen the issue, please?

**Comment by ozturkkl** [comment]
Both RTCPeerConnection and using <canvas> with MediaStreamTrackProcessor add overhead and reduce the FPS of the resulting recording. Is there any other solution that I'm not aware of? This is proving to be a massive pain point. I just want to be able to record a camera and if the feed drops, make sure it gets picked up again by re-grabbing the user media and adding the track to the existing recorder.

**Comment by EddyVinck** [comment]
I just want to be able to switch audio input device while recording in the browser

is that too much to ask for?

**Comment by alvestrand** [comment]
@kalido "no developer interest" referred to "no comment for four years". There's a limit to how long we should keep feature requests open just in case someone's interested.

@EddyVinck for your use case, what happens if you do

stream.removeTrack(oldTrack);
stream.addTrack(newTrack);

on a `stream` being recorded?

**Comment by Kaiido** [comment]
> (at)ka[i]ido "no developer interest" referred to "no comment for four years". There's a limit to how long we should keep feature requests open just in case someone's interested.

Sure if you take more than 1 year to answer every comment you won't get much... 
Latest comments in the thread by implementers seemed to say there isn't really anything preventing this from happening. What more do you want from devs? More use cases?

> what happens if you do [...]

The MediaStream Recording specs (isn't it this repo?) state

> If at any point, a track is added to or removed from stream’s [track set](https://w3c.github.io/mediacapture-main/#dfn-track-set), the UA MUST stop gathering data, and queue a task, using the DOM manipulation task source, that runs the following steps:

**Comment by alvestrand** [comment]
Other features have been much more aggressively pursued by their proponents.
Yes, I want continued engagement in order to get a feature implemented. We had a discussion on this issue from May to September 2019, and after that there was crickets.

I didn't ask what the spec currently says; I asked what happened if you tried. If some browser actually does something that the implementors like better than the spec here, we can adjust the specs to align with the implementations.

**Comment by Kaiido** [comment]
> I didn't ask what the spec currently says; I asked what happened if you tried. If some browser actually does something that the implementors like better than the spec here, we can adjust the specs to align with the implementations.

No they follow the specs here, somehow. It's [tested by wpt](https://wpt.fyi/results/mediacapture-record/MediaRecorder-error.html?label=experimental&label=master&aligned&q=mediarecorder) FF and Safari fail on the type of event, but they do fire one named "error" and stop recording.
